ntoskrnl/io/iomgr/driver.c

```diff
@@ -524,7 +524,18 @@ IopInitializeDriverModule(
     RtlZeroMemory(driverObject, ObjectSize);
     driverObject->Type = IO_TYPE_DRIVER;
     driverObject->Size = sizeof(DRIVER_OBJECT);
-    driverObject->Flags = DRVO_LEGACY_DRIVER; // TODO: check the WDM_DRIVER flag on the module
+
+    PIMAGE_NT_HEADERS NtHeaders = RtlImageNtHeader(ModuleObject->DllBase);
+    // NOTE: We suppose that since the driver has been successfully loaded,
+    // its NT and optional headers are all valid and have expected sizes.
+    ASSERT(NtHeaders);
+    ASSERT(ModuleObject->SizeOfImage == NtHeaders->OptionalHeader.SizeOfImage);
+    ASSERT(ModuleObject->EntryPoint == RVA(ModuleObject->DllBase, NtHeaders->OptionalHeader.AddressOfEntryPoint));
+
+    /* Set the legacy flag if this is not a WDM driver */
+    if (!(NtHeaders->OptionalHeader.DllCharacteristics & IMAGE_DLLCHARACTERISTICS_WDM_DRIVER))
```

Comment:
The interesting thing is, we don't put this flag on our drivers at all

Reply:
> The interesting thing is, we don't put this flag on our drivers at all

So the observed result should practically be the same, since for these the flag would still be set (as it currently was, unconditionally)

Reply:
Yes, the flag is removed below atm